src/main/java/org/terasology/alterationEffects/regenerate/RegenerationAlterationEffect.java

```diff
@@ -1,138 +1,74 @@
-        }
-        // Otherwise, if the duration is greater than 0, there are no modifiers found, and the effect modify event was not consumed,
-        // add a delayed action to the DelayManager using the old system.
-        else if (duration > 0 && !modifiersFound && !effectModifyEvent.isConsumed()) {
-            delayManager.addDelayedAction(entity, AlterationEffects.EXPIRE_TRIGGER_PREFIX + AlterationEffects.REGENERATION, duration);
-        }
-        // Otherwise, if there are either no modifiers found, or none of the modifiers collected in the event have infinite
-        // duration, remove the component associated with this regeneration effect.
-        else if (!modifiersFound || !effectModifyEvent.getHasInfDuration()) {
-            entity.removeComponent(RegenerationComponent.class);
-        }
-        // If this point is reached and none of the above if-clauses were met, that means there was at least one modifier
-        // collected in the event which has infinite duration.
+    @Override
+    protected RegenerationComponent updateComponent(OnEffectModifyEvent event, RegenerationComponent component,
+                                                    final EffectContext context) {
+        component.regenerationAmount = (int) event.getMagnitudeResultValue();
+        return component;
     }
 
     /**
-     * This will apply the regeneration effect on the given entity by calling the method
-     * {@link #applyEffect(EntityRef, EntityRef, float, long)}.
+     * This will apply the regeneration effect on the given entity by calling the method {@link #applyEffect(EntityRef,
+     * EntityRef, float, long)}.
      *
-     * @param instigator    The entity who applied the regen effect.
-     * @param entity        The entity that the regen effect is being applied on.
-     * @param id            Id is used for unique identification in regen scheduler.
-     * @param magnitude     The magnitude of the regen effect.
-     * @param duration      The duration of the regen effect.
+     * @param instigator The entity who applied the regen effect.
+     * @param entity The entity that the regen effect is being applied on.
+     * @param id Id is used for unique identification in regen scheduler.
+     * @param magnitude The magnitude of the regen effect.
+     * @param duration The duration of the regen effect.
      */
     @Override
     public void applyEffect(EntityRef instigator, EntityRef entity, String id, float magnitude, long duration) {
+        //TODO: this is supposed to be the "new" implementation with the approach above being deprecated
         //applyEffect(instigator, entity, magnitude, duration);
         if (magnitude != 0) {
+            //FIXME(skaldarnar): "duration" may be -1 to indicated DURATION_INDEFINITE - how does that translate to the
+            //                   regen event we are sending here?
```

Comment:
For a potion to regen until full health, the duration has to be -1000 (ms)? Seems like a not so ideal solution when seen from implementer (potions, for example) point of view, for using this particular number. 

It would be better if it is added in the doc for this effect that the duration is in milliseconds and this special case of regen until full health.

Reply:
Yeah, I agree that this implementation is not ideal. I plan to do an overhaul and unification of existing effects, will take this into account then.